draft-ietf-acme-acme.md

```diff
@@ -1000,6 +1000,120 @@ name validation.
 }
 ~~~~~~~~~~
 
+### Status Changes
+
+Each ACME object type goes through a simple state machine over its
+lifetime.  The "status" field of the object indicates which state
+the object is currently in.
+
+Challenge objects are created in the "pending" state.  They
+transition to the "processing" state when the client responds to the
+challenge (see {{responding-to-challenges}}) and the server begins
+attempting to validate that the client has completed the challenge.
+Note that within the "processing" state, the server may attempt to
+validate the challenge multiple times (see {{retrying-challenges}}).
+Likewise, client requests for retries do not cause a state change.
+If validation is successful, the challenge moves to the "valid"
+state; if there is an error, the challenge moves to the "invalid"
+state.
+
+~~~~~~~~~~
+         pending
+            | 
+            | Receive
+            | response
+            V
+        processing <-+
+            |   |    | Server retry or
+            |   |    | client retry request
+            |   +----+
+            |
+            |
+Successful  |   Failed
+validation  |   validation
+  +---------+---------+
+  |                   |
+  V                   V
+valid              invalid
+~~~~~~~~~~
+
+Authorization objects are created in the "pending" state.  If one of
+the challenges listed in the authorization transitions to the
+"valid" state, then the authorization also changes to the "valid"
+state.  If there is an error while the authorization is still
+pending (e.g., the authorization times out), then the authorizatoin
+transitions to the "invalid" state.  Once the authorization is in
+the valid state, it can expire ("invalid"), be deactivated by the
+client ("deactivated", see {{deactivating-an-authorization}}),
+revoked by the server ("revoked").
+
+~~~~~~~~~~
+          pending
+             |
+Error        |  Challenge -> valid
+   +---------+---------+
+   |                   |
+   V                   V
+invalid <----------- valid
+            Time       |
+           after       |
+         "expires"     |
+                       |
+           +-----------+-----------+
+           |                       |
+           |                       |
+    Client |                Server |
+   deactiv.|                revoke |
+           V                       V
+      deactivated               revoked
+~~~~~~~~~~
+
+Order objects are created in the "pending" state.  Once all of the
+authorizations listed in the order object are in the "valid" state,
+the order transitions to the "ready" state.  The order moves to the
+"processing" state after the client submits a request to the order's
+"finalize" URL and the CA begins the issuance process for the
+certificate.  Once the certificate is issued, the order enters the
+"valid" state.  If a fatal error occurs at any of these stages, the
+order moves to the "invalid" state.
```

Comment:
This paragraph (& the flow chart) are also missing three ways that an order can become invalid:
1) One of its authorizations becomes invalid because it was deactivated
2) One of its authorizations becomes invalid because it expired
3) The order expires

Reply:
> One of its authorizations becomes invalid because it was deactivated

Or the authorization was revoked. :-)

We should probably say "If any of the authorizations changes to a state other than pending or valid, the order becomes invalid."